### DISCUSSION TITLE
Duplicate Project

### DISCUSSION THREAD

**Discussion by Zegoverno** [supabase/supabase#5930] [Feature Requests] [answered]
### If you still need an easy way to duplicate your project please drop a comment below describing your use case, what exactly you wish to clone from your existing database (e.g. only schemas, schemas and all data, etc.), and any Supabase products you will continue using after duplicating your project.

---

Would be awesome to duplicate projects and also choose to copy just the database schemas/structure or also migrate all data.

**Comment by burggraf** [comment]
You can duplicate a project using the instructions here:  [~~Migrate the Database~~](https://supabase.com/docs/guides/database#migrate-the-database) edit: [new Migration docs](https://supabase.com/docs/guides/platform/migrating-within-supabase)

If you want to only duplicate the schema and not the data, you can use `pg_dump -s postgres`.

**Reply by phillipmohr** [reply]
It seems like the docs have changed... I was able to duplicate my project with the help of the previous instructions, but the step-by-step list is now gone as it seems. The docs don't seem to explain how to duplicate a project anymore, if I am not mistaken. Plus, I don't know how to access the "Table View" which is mentioned in the docs and I'm also not sure if it will help me to duplicate my entire project...

**Reply by GaryAustin1** [reply, marked as answer]
The migration method is here now:  https://supabase.com/docs/guides/platform/migrating-within-supabase
I think it is different than before, and have not tried this method.

**Comment by ghost** [comment]
I tried [https://supabase.com/docs/guides/database/migrating-between-projects](https://supabase.com/docs/guides/database/migrating-between-projects) but it is deleting everything :(

I would like to duplicate my production project to a staging project (same data, same storage).

Is it achievable ?

**Reply by awalias** [reply]
new doc: https://supabase.com/docs/guides/platform/migrating-within-supabase

**Comment by tonyneel** [comment]
I would like to create a staging environment as well but am worried by @paulin-crtn comment. Any guidance?

**Comment by logemann** [comment]
would also like to see a "duplicate" project button in the supabase UI. Would most likely also drive sales.

**Comment by tarkanlar** [comment]
That would be an amazing feature

**Comment by cerpinn** [comment]
this would be super helpful!

**Comment by danarchos** [comment]
To be able to easily create new environment, one for testing, one for development and one for production would be a dream

**Comment by 99power** [comment]
Yeah this is so necessary. Coming from mysql where you can just import/export and you're done, this feels really cumbersome.

**Comment by sjmog** [comment]
Agree – this feels super-necessary (see e.g. Render for how it could be handled)

**Comment by johndpope** [comment]
Given the in the face alerts for RLS / I need this - don't want to spend a day recreating environment. 
I'm even a paying customer. I have an older snapshot - but it's now out of sync - was shut down for long time. 
if you could make easier here to sync to an existing live database - that would also suffice. 

UPDATE -

wait a second - we have this
https://supabase.com/docs/guides/integrations/snaplet


UPDATE  2 - can you guys buy this snaplet product / company ?/ it should be out of the box.

**Comment by adhambadr** [comment]
So tried snaplet, and it messed up all the foreign key relationships, practically any foreign table query is 500 error. Its really tedious to clone a project, how are we supposed to maintain a local and live environments in sync ?!

(edited by @MildTomato to remove profanity)

**Reply by caseybaggz** [reply]
Same. I had to literally make my back-up DB production and now have to figure out how to duplicate my (now) production to a new backup...

**Comment by jmcannon** [comment]
I'd also love a "Copy this Project" button.

**Comment by Angus-Moore-Dev** [comment]
I think it would also be cool to have updates be able to be made to different projects (based on their version (say one project has a development, staging and production environment)), similar to how Github handles pull requests.

Say I make some relationship / schema changes to a few tables in development or how the storage bucket RLS works, I would like to be able to merge all of those changes from development into the production database. 

Not sure how that would realistically work, just pie in the sky thinking ¯\_(ツ)_/¯

**Comment by peachp** [comment]
Neither Snaplet nor `supabase db dump` works with our data. Snaplet fails at creating a snapshot and the supabase CLI dump creates broken SQL INSERTs which cannot be restore via psql or `supabase db reset`

**Reply by peachp** [reply]
To clarify, our data is not the cleanest (long TEXT values with all sorts of characters), but it worked just fine in legacy system and initially moving it to Supabase using the official JS client worked without issues.

**Comment by slawton3** [comment]
Anyone know if a "Duplicate project" button/feature is in the pipeline for a future release? I would instantly and gladly pay for this feature.

**Reply by saltcod** [reply]
Hey there! Some questions to help guide the way here.

How would you see this working? 
What is the pain point this would solve? Does branching solve any of this pain?
Why do you need it? What would be the benefit over just using [pg_dump as Mark has suggested](https://github.com/orgs/supabase/discussions/5930#discussioncomment-2371176)?

**Reply by Gregg-Casey** [reply]
@saltcod please read the thread above. All of your questions are answered there.

**Comment by didymu5** [comment]
I'd like this feature as well!

**Reply by saltcod** [reply]
Hey there! Some questions to help guide the way here.

How would you see this working? 
What is the pain point this would solve? Does branching solve any of this pain?
Why do you need it? What would be the benefit over just using [pg_dump as Mark has suggested](https://github.com/orgs/supabase/discussions/5930#discussioncomment-2371176)?

**Comment by alexrabin** [comment]
+1

**Comment by shane-downes** [comment]
Support for prod and dev environments would be awesome.

**Comment by akrv** [comment]
need.

for now: after few copy paste, worked great and then setup hooks and triggers manually

https://supabase.com/docs/guides/platform/migrating-and-upgrading-projects

**Comment by Shawns2759** [comment]
plz

**Comment by pooriaarab** [comment]
is this feature being added @burggraf

**Reply by saltcod** [reply]
Hey there! Some questions to help guide the way here.

How would you see this working? 
What is the pain point this would solve? Does branching solve any of this pain?
Why do you need it? What would be the benefit over just using [pg_dump as Mark has suggested](https://github.com/orgs/supabase/discussions/5930#discussioncomment-2371176)?

**Comment by zozoheir** [comment]
Also adding to this. I find it unbelievable the amount of time saved by Supabase, but then wasted b/c the most common practice (dev/prod/staging envs) is not supported. Cmon...Im not a paying user but I would happily pay for this.

**Reply by saltcod** [reply]
Hey there! Some questions to help guide the way here.

How would you see this working? 
What is the pain point this would solve? Does branching solve any of this pain?
Why do you need it? What would be the benefit over just using [pg_dump as Mark has suggested](https://github.com/orgs/supabase/discussions/5930#discussioncomment-2371176)?

**Comment by MildTomato** [comment]
> Would be awesome to duplicate projects and also choose to copy just the database schemas/structure or also migrate all data.

Update from the Supabase team — We are looking into this. We realize this feature request has been here for a while and it has gathered a lot of votes.

Due the complexity of amount of services, each with their own config, data, deployments, we can't offer in the immediate short term a service to duplicate _everything_ in a Supabase project just yet. so we'll think of this as a medium to long term goal.

However, that shouldn't stop us making at least some progress that can satisfy some use cases. So we can start with duplicating the database schema, and we'll aim to roll that out fairly quickly as an easy action you can take in the Dashboard (and probably CLI).

Although we can't promise any timeframes, Database data duplication is probably something we'll consider next, and then we'll start tackling the issues with Storage duplication, Function deployments and so on.

**Reply by jzambranoavvy** [reply]
Let's goooooo!

**Comment by logemann** [comment]
Its ok now.


On Sun, Oct 15, 2023 at 9:39 AM Jonathan Summers-Muir <
***@***.***> wrote:

> Would be awesome to duplicate projects and also choose to copy just the
> database schemas/structure or also migrate all data.
>
> Update from the Supabase team — We are looking into this. We realize this
> feature request has been here for a while and it has gathered a lot of
> votes.
>
> Due the complexity of amount of services, each with their own config,
> data, deployments, we can't offer in the immediate short term a service to
> duplicate *everything* in a Supabase project just yet. so we'll think of
> this as a medium to long term goal.
>
> However, that shouldn't stop us making at least some progress that can
> satisfy some use cases. So we can start with duplicating the database
> schema, and we'll aim to roll that out fairly quickly as an easy action you
> can take in the Dashboard (and probably CLI).
>
> Although we can't promise any timeframes, Database data duplication is
> probably something we'll consider next, and then we'll start tackling the
> issues with Storage duplication, Function deployments and so on.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/orgs/supabase/discussions/5930#discussioncomment-7283892>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACUR4PL4VJI54AJZGIPCQLX7OHMPAVCNFSM5QZQVD3KU5DIOJSWCZC7NNSXTOKENFZWG5LTONUW63SDN5WW2ZLOOQ5TOMRYGM4DSMQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
regards
marc logemann
https://www.linkedin.com/in/logemann/

**Comment by wiiiimm** [comment]
I think Supabase's upcoming [Branching](https://supabase.com/blog/supabase-local-dev#branching-and-preview-environments) feature will help to address this.

**Reply by Gregg-Casey** [reply]
The branches aren't meant to be permanent so I am not sure about that @williamli .

**Reply by tarekxsellami** [reply]
a branch costs money each day

**Comment by mounikpatel** [comment]
I am also looking for a similar implementation where I can copy my whole database (minus the data) from dev env to prod env. Both environments are on different user accounts as well. I would really appreciate any help.

**Reply by gwc4github** [reply]
it would be great to be able to select individual tables to migrate the data as well. there are some meditated tables that have the same data in all environments.

**Comment by alramalho** [comment]
same here, commenting for support 😅

**Comment by modbender** [comment]
What I want is a duplicate of schema on something like dev and prod, where schema changes can be safely pushed from dev to prod. Only data will differ.

**Comment by hixcoder** [comment]
for me I used two methods:
[I'm using windows]

### **method 1:**
1- you can install [postgres ](https://www.postgresql.org/download/).
2- dd the path to the PostgreSQL bin directory to your system’s PATH. This directory is typically C:\Program Files\PostgreSQL\<version>\bin.
3- open a command prompt and type `pg_dump --version` to verify that it was installed correctly.
4- export your supabase project to your machine:
`pg_dump -h <old_project_db_host> -U <old_project_db_user> -d <old_project_db_name> -p <old_project_db_port> -Fc > full_backup.dump`
5-import it to a new project:
`pg_restore -h <new_project_db_host> -U <new_project_db_user> -d <new_project_db_name> -p <new_project_db_port> -Fc full_backup.dump`

_Note: this method will duplicate everything , except the storage_

### **method 2:**
1- use this documentation [here](https://supabase.com/docs/guides/platform/migrating-and-upgrading-projects#migrate-your-project).

_Note: this method will duplicate everything!_

**Reply by erpview** [reply]
well ive created app that makes life esiar first

It checks do you have installed Supabase CLi, Docker and PosgreSQL min 17.2.

If yes than ask you for Source Database URL and Target Database URL (old project and new project). Test conection and make full dump to new database. For me work. If someone need ill live link to mac installer and windows
<img width="1179" alt="Zrzut ekranu 2024-12-7 o 00 47 56" src="https://github.com/user-attachments/assets/b4c60629-de1e-4596-a516-c2b21e2ec6ed">
<img width="1186" alt="Zrzut ekranu 2024-12-7 o 00 48 07" src="https://github.com/user-attachments/assets/181659b2-f3dc-4c4e-9a4d-2c2c6e1f7fc1">
